consensus/src/checkpoint.rs

```diff
@@ -112,47 +117,221 @@ mod tests {
         );
     }
 
+    fn gen_num_blocks_in_epochs(slot_fill_rate: f64, slot_fill_rate_delta: f64, n: f64) -> i32 {
+        let protocol_constants = ProtocolConstantsTemp::default();
+        let nums = gen_num_blocks_in_slots(
+            slot_fill_rate,
+            slot_fill_rate_delta,
+            n * protocol_constants.slots_per_epoch.0 as f64,
+        );
+        nums
+    }
+
+    fn gen_spot_root_epoch_position(slot_fill_rate: f64, slot_fill_rate_delta: f64) -> (u32, u32) {
+        //  We need to simulate both the staking epoch and the next staking epoch,
+        //  the root epoch is the staking epoch. The root epoch position this function generates
+        //   is the epoch number of the staking epoch and the block height the
+        //  staking epoch starts at (the simulation of all blocks preceeding the
+        //  staking epoch
+        let root_epoch_int = thread_rng().gen_range(0..100);
+        let root_block_height =
+            gen_num_blocks_in_epochs(slot_fill_rate, slot_fill_rate_delta, root_epoch_int as f64);
+        return (root_epoch_int, root_block_height as u32);
+    }
+
+    fn convert(x: f64) -> i32 {
+        x.round().rem_euclid(2f64.powi(32)) as u32 as i32
```

Comment:
Perf: Define constant for `2f64.powi(32)`

Reply:
A const function cannot call a non-const function see [link](https://stackoverflow.com/questions/45249486/why-cant-const-functions-in-rust-make-calls-to-associated-functions)
![Screenshot (389)](https://user-images.githubusercontent.com/8589224/148331427-9f660377-719e-40d5-8a80-85b2d61d2ff4.png)

Reply:
You can do either
```rust
const a:f64 = 2_u64.pow(32) as f64;
```
or
```rust
const a:f64 = 4294967296_f64
```

Reply:
Thank you for the excellent guidance 🥇